File: Project_and_Analysis_of_Algorithms/graph_search_applications/c++/toposort.cpp
#include "graphs.h"

// Topological sort for DAGs using BGS (Kahn algorithm).
vector<string> topoSortKahn(adjacencyMap(string) g) {
    // Input degrees.
    map<string, int> inDeg; for(auto const&[node, _] : g) inDeg[node] = 0;
    
    // Node inputs at neighbour.
    for(auto const&[node, _] : g) {
        for (string neighbour : g[node]) inDeg[neighbour] += 1;
    }
    // Source nodes list.
    queue<string> q;
    
    //Output nodes order.
    vector<string> order;
    
    // Inits q with nodes with input degree equals 0.
    for(auto const&[node, _] : g) {
        if (inDeg[node] == 0) q.push(node);
    }

    while(!q.empty()) {
        string curNode = q.front(); q.pop();
        order.push_back(curNode);
        for (string neighbour : g[curNode]) {
            inDeg[neighbour]--;
            // If input degree is 0, it will be a source node.
            if(inDeg[neighbour] == 0) q.push(neighbour);
        }
    }

    return order;
}

void dfsRec(adjacencyMap(string) g, string startNode, visitedMap(string) &visited, vector<string> &order) {
    visited[startNode] = true;
    for (string neighbour : g[startNode]) {
        if (!visited[neighbour]) {
            dfsRec(g, neighbour, visited, order);
        }
    }
    order.push_back(startNode);
}

// Topological sorting using DFS.
vector<string> dfsTopoSort(adjacencyMap(string) g){
    visitedMap(string) visited = initVisited(g);
    vector<string> order = {};

    for (auto const&[node, _] : g) {
        if (!visited[node]) {
            dfsRec(g, node, visited, order);
        }
    }

    reverse(order.begin(), order.end());
    return order;
}

int main(void) {
    for (string node : topoSortKahn(G1)) cout << node << "\n";
    cout << "--------------\n";
    for (string node : dfsTopoSort(G1)) cout << node << "\n";
    return EXIT_SUCCESS;
}

File: Project_and_Analysis_of_Algorithms/graph_search_applications/python/toposort.py
from graphs import G1, G4, init_visited
from collections import deque

def toposort_kahn(g: dict[str, list[str]]) -> list[str]:
    """Topological sort for DAGs using BGS (Kahn algorithm)."""
    in_deg: dict[str, int] = {node: 0 for node in g.keys()} # Input degrees.

    for node in g.keys():
        for neighbour in g[node]:
            in_deg[neighbour] += 1 # Node inputs at neighbour.

    q: deque[str] = deque() # Source nodes list.
    order: list[str] = [] # Output nodes order.

    for node in g.keys():
        if in_deg[node] == 0:
            q.append(node) # Inits q with nodes with input degree equals 0.

    while len(q) > 0:
        cur_node: str = q.popleft() 
        order.append(cur_node)
        for neighbour in g[cur_node]:
            in_deg[neighbour] -= 1
            if in_deg[neighbour] == 0: # If input degree is 0, it will be a source node.
                q.append(neighbour)

    return order


def dfs_toposort(g: dict[str, list[str]]) -> list[str]:
    """DFS topological sort."""
    visited: dict[str, bool] = init_visited(g)
    order: list[str] = []

    def dfs_rec(g: dict[str, list[str]], start_node: str, visited: dict[str, bool], order: list[str]):
        visited[start_node] = True
        for neighbour in g[start_node]:
            if not visited[neighbour]:
                dfs_rec(g, neighbour, visited, order)
        order.append(start_node)

    for node in g.keys():
        if not visited[node]:
            dfs_rec(g, node, visited, order)

    order.reverse()
    return order


def main():
    print(toposort_kahn(G1))
    print("--------------")
    print(dfs_toposort(G4))


if __name__ == "__main__":
    main()
